hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java

```diff
@@ -518,14 +475,31 @@ private ContainerMoveSelection matchSourceWithTarget(
       }
       return null;
     }
+
+    //if the utilization of the source data node becomes lower than lowerLimit
+    //after the container is moved out , then the container can not be
+    // a candidate one, and we should remove it from the candidateContainers.
+    candidateContainers.removeIf(c -> {
+      ContainerInfo cInfo;
+      try {
+        cInfo = containerManager.getContainer(c);
+      } catch (ContainerNotFoundException e) {
+        LOG.warn("Could not find container {} when " +
+            "be matched with a move target", c);
+        //remove this not found container
+        return true;
+      }
+      return !findSourceStrategy.canSizeLeaveSource(
+          source, cInfo.getUsedBytes());
+    });
```

Comment:
Can we move this to selection criteria#getCandidateContainers? Maybe selection criteria can have reference to find source strategy.

Reply:
looks good , will do this, thanks!